### PR TITLE
feat(topk): add varlen tie-break and page-table parity

### PR DESCRIPTION
**Pull request by zianglih**
## 📌 Description

@humansand

This draft rebases the varlen Top-K work onto merged #4315, adds exact boundary tie semantics to the existing CuTe radix implementation, and fuses compact page translation into that same native kernel.

- **Tie contract:** `TopKTieBreak.SMALL`/`LARGE` selects the smaller/larger local indices only among values equal to the K-th value. It changes the selected set, not output order.
- **Incremental kernel change:** following #3095, the existing single-/multi-CTA radix collection phase is specialized at compile time. `NONE` retains the original atomic path; tie modes reuse the radix pivot, rank pivot-equal candidates across CTA chunks, and use the existing collection machinery. No standalone selector kernel was added.
- **Native score windows:** `row_starts` now shifts radix loads while returned indices remain window-local. Effective lengths are clamped to `[0, row_width - row_start]`, including empty/negative speculative lengths and oversized `seq_lens`; short-row value padding is zero.
- **Native compact pages:** `top_k_varlen_page_table_transform` extends the same radix kernel with `src_page_table`, `row_to_batch`, `row_starts`, `page_table_row_starts`, `page_size`, `out`, `out_values`, and `out_raw_indices`. Physical translation happens after the existing collection phase, with no second selector kernel and no call or fallback to the earlier C++/CUTLASS page-table implementation.
- **API changes:** varlen paths are graph safe after one-time JIT/module warmup, so `dsa_graph_safe` was removed. `deterministic` remains in both signatures for forward compatibility, but `deterministic=True` raises `ValueError` before dispatch; it never selects another backend.
- **Routing:** native `row_starts` and compact-page calls use CuTe radix on Blackwell-plus. The regular non-page API retains `radix_cutlass` as its graph-safe fallback where supported; the compact-page sibling is radix-only and reports unsupported hardware instead of falling back.
- **GVR boundary:** GVR remains `TopKTieBreak.NONE`-only. Its bounded candidate buffer can discard members of an arbitrarily large K-th-value tie before final collection, so SMALL/LARGE cannot be made exact only by changing writeback. Explicit GVR tie requests fail, while `auto` selects radix for tie modes on Blackwell.
- **Benchmarks/trace:** the unified benchmark now uses CUDA-graph timing for `top_k_varlen`; `bench_topk.py` includes `top_k_varlen_page_table/radix` rows and tie variants. The page trace and generated JSON now target native radix and contain no `dsa_graph_safe` input.
- **Stack status:** based directly on merged #4315 at `7f614b86470180bab2d22e36fd1775791c6bf3e6`; tested and pushed PR head is `f81c30caa94c279618c6393072c3478df423153b`.

## 🔍 Related Issues

- Rebased onto the merged compact page-table work in #4315.
- Follows the boundary-set contract and incremental-kernel pattern from #3095 and #3133.
- Extends the varlen Top-K API introduced in #3901.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit`.
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run `pre-commit run --all-files` and fixed reported issues.

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] CUDA tests are passing.

### Environment

- **Devbox:** `flashinfer-pr4421-cu132`, HAI cluster `c1`, namespace `infra`.
- **Image:** `flashinfer/flashinfer-ci-cu132:20260408-4cce866`.
- **Hardware:** 8× NVIDIA B200 exposed, driver `580.126.09`; validation and benchmarks used CUDA device 0.
- **Toolchain:** CUDA compiler `13.2.51`, PyTorch `2.12.0.dev20260408+cu132`, FlashInfer `0.6.18`, CUTLASS Python `4.7.0`.
- **Timing packages:** `cupti-python==13.3.1`, `nvidia-cuda-cupti==13.3.75`.
- **Timing setup:** `python -m pip install -U cupti-python` in the CUDA 13.2 image.
- **Environment variables:** none required. The unified benchmark uses CUPTI and CUDA graphs by default for this routine.

Devbox request:

```bash
uv run h devbox flashinfer-pr4421-cu132 \
  -c c1 -n infra -g b200 --gpus-per-pod 8 --size xlarge \
  --image flashinfer/flashinfer-ci-cu132:20260408-4cce866 --bare
```

### Static validation

```bash
uvx pre-commit run --all-files
git diff --check
```

```text
pre-commit: all hooks passed
git diff --check: exit 0
```

### CUDA correctness

```bash
pytest -q tests/topk_varlen
pytest -q -rs \
  tests/trace/test_top_k_varlen_reference_correctness.py \
  tests/trace/test_fi_trace_template_consistency.py \
  tests/trace/test_template_init.py \
  -k top_k_varlen
```

```text
tests/topk_varlen:
148 passed, 621 warnings in 57.28s

focused trace/reference/template coverage:
16 passed, 2 skipped, 972 deselected, 3 warnings in 0.51s
SKIPPED [1] tests/trace/test_template_init.py:273: [top_k_varlen] no paged-KV arrays in init result
SKIPPED [1] tests/trace/test_template_init.py:273: [top_k_varlen_page_table_transform] no paged-KV arrays in init result
```

Coverage includes exact SMALL/LARGE boundary sets in single- and multi-CTA radix, native score-window/value semantics, speculative empty lengths, page sizes 1/4/64, independent score/page starts, default and explicit row-to-batch mappings, preallocated raw/physical/value outputs, alias rejection, and compact-page CUDA-graph replay after mutating logits and mapping tensors. The committed trace artifact is `tests/trace/fi_trace_out/top_k_varlen_page_table_transform_n8192_k1024_ps64.json`.

#### Native radix ties versus the regular CUTLASS fallback

```bash
python benchmarks/flashinfer_benchmark.py \
  --routine top_k_varlen \
  --batch_size 16 --max_seq_len 65536 --top_k 512 \
  --input_dtype bfloat16 \
  --backends radix radix_cutlass \
  --tie_break all \
  --num_iters 30 --dry_run_iters 5 --refcheck
```

- **Workload:** BF16 logits, `B=16`, `N=65,536`, `K=512`, full per-row lengths, `next_n=1`, `compress_ratio=1`, indices only.
- **Timing:** CUPTI device activity timing; median of 30 captured iterations after 5 dry runs. Lower is better.
- **Traffic:** `B*N*sizeof(BF16) + B*sizeof(int32 seq_len) + B*K*sizeof(int32 output)` divided by median time. GVR is excluded because SMALL/LARGE is unsupported.

```text
[PERF] radix          :: median time 0.015 ms; std 0.000 ms; achieved tflops 0.000 TFLOPs/sec; achieved tb_per_sec 0.139 TB/sec
[PERF] radix/tie_small:: median time 0.020 ms; std 0.000 ms; achieved tflops 0.000 TFLOPs/sec; achieved tb_per_sec 0.105 TB/sec
[PERF] radix/tie_large:: median time 0.020 ms; std 0.000 ms; achieved tflops 0.000 TFLOPs/sec; achieved tb_per_sec 0.105 TB/sec
[PERF] radix_cutlass  :: median time 0.045 ms; std 0.000 ms; achieved tflops 0.000 TFLOPs/sec; achieved tb_per_sec 0.048 TB/sec
[PERF] radix_cutlass/tie_small:: median time 0.058 ms; std 0.000 ms; achieved tflops 0.000 TFLOPs/sec; achieved tb_per_sec 0.037 TB/sec
[PERF] radix_cutlass/tie_large:: median time 0.060 ms; std 0.000 ms; achieved tflops 0.000 TFLOPs/sec; achieved tb_per_sec 0.036 TB/sec
```

Derived summary: native SMALL/LARGE add about 33% over radix `NONE` and remain about 2.9×/3.0× faster than the corresponding regular CUTLASS fallback modes.

#### Fused native compact-page stores

```bash
python - <<'PY'
import torch
from benchmarks.bench_topk import VarLenCase, bench_top_k_varlen_backend

case = VarLenCase(
    name="decode_uniform_b16_l65536_k512",
    regime="decode",
    length_dist="uniform",
    num_requests=16,
    q_len=1,
    max_len=65536,
    k=512,
)
generator = torch.Generator(device="cuda")
generator.manual_seed(1234)
result = bench_top_k_varlen_backend(
    case,
    "radix",
    torch.bfloat16,
    generator,
    compare_tie_break=True,
    page_table=True,
)
for key in (
    "flashinfer_us",
    "flashinfer_tie_small_us",
    "flashinfer_tie_large_us",
    "tie_small_slowdown_vs_baseline",
    "tie_large_slowdown_vs_baseline",
):
    print(key, result[key])
PY
```

- **Workload:** seed-1234 uniformly sampled decode lengths, BF16 `B=16`, `N=65,536`, `K=512`, compact `page_size=64`, physical indices only, preallocated output.
- **Timing:** CUPTI median of 100 iterations after 10 dry runs; CUDA graph disabled in this comparison helper. Lower is better.

```text
flashinfer_us 15.920499999999997
flashinfer_tie_small_us 21.12
flashinfer_tie_large_us 21.264
tie_small_slowdown_vs_baseline 1.3265915015231935
tie_large_slowdown_vs_baseline 1.3356364435790335
```

### Limitations and untested scope

- GVR SMALL/LARGE is deliberately unsupported because its bounded candidate buffer cannot preserve an arbitrary K-th-value tie; GVR `NONE` is unchanged.
- `deterministic=True` is deliberately unsupported for both varlen APIs and does not fall back.
- Native score windows and the compact-page sibling require Blackwell-plus/CuTe DSL. No non-Blackwell GPU was available; compact-page calls on unsupported hardware do not fall back to C++.
- Page translation with tie modes was benchmarked but not separately captured in a CUDA graph. The suite covers captured native page translation for `NONE`, captured native radix, and existing GVR graph paths; the unified benchmark captures regular radix tie modes.
- Multi-GPU/distributed execution was not tested; all CUDA validation used device 0.
- The two trace-template skips are expected because these Top-K initializers do not return paged-KV cache arrays. Existing CuTe DSL deprecation warnings remain; there were no test failures.

## Reviewer Notes

- Review selected-set semantics independently from output ordering; `deterministic=True` is not implemented.
- Review the compile-time `NONE`/`SMALL`/`LARGE` collection specializations and global pivot-equality ranking in `radix_topk.py`.
- Review score-window clamping and the post-collection compact-page translation/barriers in the existing radix kernel.
- Review the native-only page routing and absence of any C++ page-table fallback.
- Review the explicit GVR tie rejection; this intentionally prefers exact semantics over an approximate candidate-capped path.

### CODE REVIEW

**Comment by zianglih** [comment]
Closing for now due to too much effort
